gum/backend-linux/gummemoryaccessmonitor-linux.c

```diff
@@ -0,0 +1,464 @@
+    GumAddress end = cur + r->size;
+
+    while ((range = g_list_next (range)))
+    {
+      GumLinuxRange * lr = (GumLinuxRange *) range->data;
+      do
+      {
+        if (cur >= lr->base &&
+            cur + self->page_size - 1 < lr->base + lr->size)
+        {
+          GumPageDetails details;
+
+          details.range.base = cur;
+          details.range.size = self->page_size;
+          details.range.prot = lr->prot;
+          details.range_index = i;
+          carry_on = func (&details, user_data);
+          cur += self->page_size;
+        }
+        else
+        {
+          break;
+        }
+
+      }
+      while (cur < end && carry_on);
+    }
+  }
+
+  g_list_free_full (head, gum_linux_free_ranges);
+}
+
+static void
+gum_linux_free_ranges (gpointer data)
+{
+  g_slice_free (GumLinuxRange, data);
+}
+
+static gboolean
+gum_linux_set_flag (const GumPageDetails * details,
+                    gpointer user_data)
+{
+  GumMemoryAccessMonitor * self = GUM_MEMORY_ACCESS_MONITOR (user_data);
+  const GumLinuxRange * range = &details->range;
+  GumPageProtection access_mask = self->access_mask;
+  GumPageProtection new_prot = GUM_PAGE_NO_ACCESS;
+  guint num_pages;
+
+  switch (range->prot)
+  {
+    case GUM_PAGE_READ:
+      if ((access_mask & GUM_PAGE_READ) != 0)
+        new_prot = GUM_PAGE_WRITE;
+      else
+        return TRUE;
+      break;
+    case GUM_PAGE_WRITE:
+      if ((access_mask & GUM_PAGE_WRITE) != 0)
+        new_prot = GUM_PAGE_READ;
+      else
+        return TRUE;
+      break;
+    case GUM_PAGE_EXECUTE:
+      if ((access_mask & GUM_PAGE_EXECUTE) != 0)
+        new_prot = GUM_PAGE_WRITE;
+      else
+        return TRUE;
+      break;
+    case GUM_PAGE_RW:
+      if (access_mask == GUM_PAGE_WRITE)
+        new_prot = GUM_PAGE_READ;
+      else if (access_mask == GUM_PAGE_READ || access_mask == GUM_PAGE_EXECUTE)
+        new_prot = GUM_PAGE_WRITE;
+      else if (access_mask == (GUM_PAGE_EXECUTE | GUM_PAGE_WRITE))
+        new_prot = GUM_PAGE_NO_ACCESS;
+      break;
+    case GUM_PAGE_RX:
+      if (access_mask == GUM_PAGE_READ ||
+          access_mask == GUM_PAGE_EXECUTE ||
+          access_mask == GUM_PAGE_RX)
+        new_prot = GUM_PAGE_WRITE;
+      else
+        return TRUE;
+      break;
+    case GUM_PAGE_RWX:
+      new_prot = GUM_PAGE_NO_ACCESS;
+      break;
+    default:
+      g_assert_not_reached ();
+  }
+
+  num_pages = self->num_pages;
+
+  self->pages_details = g_realloc (self->pages_details,
+      (num_pages + 1) * sizeof (GumPageDetails));
+
+  memcpy ((void *) &self->pages_details[num_pages],
+      (void *) details, sizeof (GumPageDetails));
+
+  self->pages_details[num_pages].completed = 0;
+  self->num_pages++;
+
+  gum_mprotect (GSIZE_TO_POINTER (range->base), range->size, new_prot);
+
+  return TRUE;
+}
+
+static gboolean
+gum_linux_clear_flag (const GumPageDetails * details,
+                      gpointer user_data)
+{
+  guint i;
```

Comment:
Variables should be ordered chronologically (except for the one holding the return value, which should be first).